### DISCUSSION TITLE
How to invite other users into the group

### DISCUSSION THREAD

**Discussion by hitohata** [genesis-tech-tribe/nishiki-documents#86] [General]
# Agenda

In this system, to share the container, the user needs to invite other users into the group. In this discussion, decide how to invite other users into the user's group.

For now, @kotaaaa suggested the following process:

Let's say user A tries to share the group with user B.

First, the Frontend accesses the '/group/{groupId}/share' API. Then it API returns Hash data that is generated by some logic within the Backend.
 -> User A gets a link address, like https://{nishiki-domain}/json/{hashValue}.
 -> User A shares that link with User B.
 -> User B accesses that link.
 -> Inside the Frontend used by User B calls Backend API, /groups/{groupId}/join/{hashValue}.
 -> Then User B joining the group is accomplished.

**Comment by hitohata** [comment]
I think @kotaaaa 's way sounds good.
I plan to create the Hash Value from Group ID and The timestamp which is a link expired Datetime.
And I'd like to make sure one thing, that we don't need to specify the user. Is this specification correct?

**Reply by nickyito** [reply]
We haven't yet talked about its specifications.
I don't think it needs to specify a user. But the number of people and days to the expiry could be added.

My idea is to set pre-defined values for both of them, for example, [ unlimited number of people ] + [ 3 days to the expiry ].

FYI: The following images are examples of how the invitation of the Discord servers works:

<img width="434" alt="image" src="https://github.com/genesis-tech-tribe/nishiki-documents/assets/99148565/d0dbce50-31f1-4930-bd39-5cbadc8df02b">
<br><br>
<img width="435" alt="image" src="https://github.com/genesis-tech-tribe/nishiki-documents/assets/99148565/32c73b33-e082-4f5c-a38f-2ca7f57cb924">

**Reply by hitohata** [reply]
> I don't think it needs to specify a user. But the number of people and days to the expiry could be added.

I don't think it needs to specify the number. Do you care that you want to prevent someone from joining the group shared via the share link?

**Reply by hitohata** [reply]
I've come to another idea. I think there are two types of inviting the group.

The first one is public, and the other one is private. In public, we don't need the limit except for expiry. Whereas, in private, it needs to specify target users.

The post data will be the following:

```json
{
  targetUsersId: [], //a list of target user's IDs
  expiryLimit: 42
}
```

* if target users are not provided, this request is treated as public.
* default expiry limit is 1 day.

By the way, the expiry limit should be changeable? I think it needs to determine the min and max limit, but it should be set free.

**Reply by nickyito** [reply]
@hitohata
If we implement the "private" invitation, what's your preferred method for searching for a user? Do you search by name or email?

**Reply by kotaaaa** [reply]
I thought we can set defalut value to 

`number of people: 1`
`expiry date: 1 day`

We can add additional feature such as selection of number of people etc, if we have any pain point after 1st release. How about this?
This is to keep minimum feature for 1st release.

**Reply by nickyito** [reply]
Yes, I agree to minimize the features of the first release.
In my opinion, the number of people could be "unlimited" because there are cases where multiple people need to be invited.

**Reply by kotaaaa** [reply]
> In my opinion, the number of people could be "unlimited" because there are cases where multiple people need to be invited.

I thought we should set default number of people to lower because in case for the share link is leaked to the outside world.

**Reply by taniguchiiqqq** [reply]
I think the expiration date should be selectable, and it can be a future improvement to limit the number of people. 
I want to clarify this. Does the number of people mean the limitation of a group member or the limitation of people who can use an issued invitation link?

**Reply by nickyito** [reply]
> Does the number of people mean the limitation of a group member or the limitation of people who can use an issued invitation link?

I was imagining the number of people that can join the group using the link.

**Reply by kanta1207** [reply]
I basically agree with @uskayyyyy on setting defaults value to both expiry date limitation of people to invite via link. It makes sense to minimize and simply that feature, especially in first release

**Reply by taniguchiiqqq** [reply]
@uskayyyyy 
Thank you for the clarification. But honestly, I don't get the point of the limitation, because all of the group members can issue the invitation link, meaning no one can't handle the limit of the number of group members. Can you explain the purpose of having the limitation? Maybe I overlooked something.

**Reply by nickyito** [reply]
@taniguchiiqqq 
> all of the group members can issue the invitation link

This is correct.


> meaning no one can't handle the limit of the number of group members

You may be misunderstanding. The "limit of the number of people" means "the number of non-members that can be invited to the group" or, in other words, "how many people can be joined using the same link". You could also rephrase it to "max number of uses of the link".

If the limit is 3 uses (= people), for example, the link becomes invalid as soon as the third new member joins the group.

Get it?

And, again, my opinion is to set the "max number of uses of the link" to "No Limit" just like how Discord works.
<img width="400" alt="image" src="https://github.com/genesis-tech-tribe/nishiki-documents/assets/99148565/5d017bb7-59fe-4ac0-85dc-da34338be3db">

**Reply by hitohata** [reply]
@uskayyyyy 
>And, again, my opinion is to set the "max number of uses of the link" to "No Limit" just like how Discord works.

Let me confirm it, do we need the maximum number of users?

**Reply by nickyito** [reply]
@hitohata 
What do you mean by the "maximum number of _users_"?

I said, "max number of _uses_"? And I think it can be "No Limit".

**Reply by hitohata** [reply]
@uskayyyyy 
I misunderstood 'user' and 'uses'. sorry.
So, I mean the number of limitations against users is not mandatory? In stage 1, should this link have a max number of users?

**Reply by nickyito** [reply]
@hitohata 
I don't think we need to limit the number of _uses_ for the first release. Meaning "No Limit", again 😂

**Reply by hitohata** [reply]
OK, let's make a decision.
I agree with @uskayyyyy. In my opinion, for stage 1, the limitation is only the expiry time and it can be selectable. (I think the limitation against users or a number of users is important, so it should be implemented after stage 1 finished.)
Does anybody have an opposite opinion or comment? If there isn't, I'd like to fix this specification.
If you have an opinion, please write a comment or push the negative reaction button (like :-1: ) until this Friday.

**Reply by kotaaaa** [reply]
I agree with @uskayyyyy and @hitohata as well for this point.
> For the first release, "No Limit" of the number of uses

But how about setting expiry time to default like `7 days` or anything. 
So that we don't have to add another screen with figma. Current invite modal (below) is enough for this feature.
<img width="436" alt="image" src="https://github.com/genesis-tech-tribe/nishiki-documents/assets/25422441/1f1fbd2d-d43c-4a47-ad0e-fe0170e4763b">
(If expiry time can be selectable, we need another screen to be designed)
I just want to focus on existing features for stage 1.

**Reply by hitohata** [reply]
@kotaaaa 
OK. The selectability is not a priority, so let's discard its feature for now.
Is the default expiry time `7 days` enough? It is just my sense but it feels too long (this time is not important for now though).

**Reply by kotaaaa** [reply]
Any number from `1 day` to `7 day` is ok for me (I suppose `3 day` is the best for this function).
@uskayyyyy  Can I get your opinion for this point?

And 1 question, will a new link be issued every time users click the `Copy Link` button? or existing links should be returned as long as they have a valid expiry time?

**Reply by hitohata** [reply]
> And 1 question, will a new link be issued every time users click the Copy Link button? or existing links should be returned as long as they have a valid expiry time?

Oh, I overlooked that situation. I think when the button is clicked, the previous one should be disabled and a new link will be created.

**Reply by kotaaaa** [reply]
> Oh, I overlooked that situation. I think when the button is clicked, the previous one should be disabled and a new link will be created.
 
If so, some user might know unexpectedly that the link is disabled when new link is issued by other user in the same group because according to the above discussion, this link will be issued uniquely by each group.
Is that permitted?


I summarized 4 candidate patterns. Please vote or comment whatever you want.
- 1, create a new link every time the button is pushed, and previous one is disabled.
- 2, create a new link every time the button is pushed, and previous one is not disabled.
- 3, don't create a new link every time the button is pushed, and share the link within the group until expiry date comes.(expiry date can be seen differently by users in the same group)
- 4, don't create a new link every time the button is pushed, and share the link within the group forever

After I summarize, I thought No3 is ok.

**Reply by hitohata** [reply]
Thank you for summarizing! You are right, considering the user experience my opinion seems a bad idea.

The No3 sounds good to me too. Is it possible to implement the expiry date for the users about the expiry date when users push the button? It might help users' experience.
However, doesn't need the updating of the expiry date? For now, users cannot expand the expiry date until that time comes. (Maybe this feature is not mandatory for now)

**Comment by nickyito** [comment]
After discussing verbally, we reached the following conclusions regarding the invitation links:
- Each link is identified by a unique hash value.
- The links expire 24 hours after they are generated.
- If a previous link is still valid, its expiry is overridden to 24 hours instead of generating a new link.
- If there is no previous valid link, a new link is generated.
